File: ObjC/RestaurantWidget/TodayViewController.m
//
//  TodayViewController.m
//  RestaurantWidget
//
//  Created by AppsFoundation on 23/09/15.
//  Copyright © 2015 AppsFoundation. All rights reserved.
//

#import "TodayViewController.h"
#import <NotificationCenter/NotificationCenter.h>
#import "MenuItem.h"
#import "MenuItemsManager.h"
#import "MenuItemTableViewCell.h"

static const CGFloat cellRatio = 0.57;

@interface TodayViewController () <NCWidgetProviding>

@property (nonatomic,strong) NSArray *menuItems;

@end

@implementation TodayViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.menuItems = [[MenuItemsManager sharedManager] loadData];
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.menuItems.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    MenuItemTableViewCell *cell = (MenuItemTableViewCell *)[tableView dequeueReusableCellWithIdentifier:@"menuItemCell"];
    MenuItem *item = self.menuItems[indexPath.row];
    
    cell.menuItemNameLabel.text = item.name;
    cell.ingredientsItemLabel.text = item.ingredients;
    cell.priceItemLabel.text = item.price;
    cell.menuItemImageView.image = [UIImage imageNamed:item.image];
    cell.discountLabel.text = item.discount;
    cell.discountLabel.hidden = item.discount == nil;
    cell.backgroundColor = [UIColor clearColor];
    
    return cell;
}

#pragma mark - UITableViewDelegate

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    CGFloat height =  tableView.bounds.size.width * cellRatio;
    return height;
}

- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath {
    self.tableViewHeightConstraint.constant += cell.bounds.size.height;
    [self.tableView needsUpdateConstraints];
}

#pragma mark - NCWidgetProviding

- (void)widgetPerformUpdateWithCompletionHandler:(void (^)(NCUpdateResult))completionHandler {
    completionHandler(NCUpdateResultNewData);
}

- (UIEdgeInsets)widgetMarginInsetsForProposedMarginInsets:(UIEdgeInsets)defaultMarginInsets {
    defaultMarginInsets.right = defaultMarginInsets.left;
    [self.view needsUpdateConstraints];
    [self.tableView needsUpdateConstraints];
    return defaultMarginInsets;
}

@end

File: Swift/RestaurantWidget/TodayViewController.swift
//
//  TodayViewController.swift
//  RestaurantWidget
//
//  Created by AppsFoundation on 23/09/15.
//  Copyright © 2015 AppsFoundation. All rights reserved.
//

import UIKit
import NotificationCenter

let CellRatio: CGFloat = 0.57

class TodayViewController: UIViewController {
    
    @IBOutlet weak var tableView: UITableView?
    @IBOutlet weak var tableViewHeightConstraint: NSLayoutConstraint?

    var menuItems: [MenuItem] = []
    
    override func viewDidLoad() {
        super.viewDidLoad()
        menuItems = MenuItemsManager.sharedManager.loadData()
    }

}

// MARK: - UITableViewDataSource
extension TodayViewController: UITableViewDataSource {
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return menuItems.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell: MenuItemTableViewCell = tableView.dequeueReusableCell(withIdentifier: "menuItemCell") as! MenuItemTableViewCell
        let item = menuItems[indexPath.row]
        
        cell.menuItemNameLabel?.text = item.name
        cell.ingredientsItemLabel?.text = item.ingredients
        cell.priceItemLabel?.text = item.price
        cell.menuItemImageView?.image = UIImage(named: item.image)
        
        if let discount = item.discount {
            cell.discountLabel?.text = discount
        } else {
            cell.discountLabel?.isHidden = true
        }
        
        return cell
    }
    
}

// MARK: - UITableViewDataSource
extension TodayViewController: UITableViewDelegate {
    
    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        tableViewHeightConstraint!.constant += cell.bounds.size.height;
        self.tableView!.needsUpdateConstraints()
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return tableView.bounds.size.width * CellRatio;
    }

}

// MARK: - NCWidgetProviding
extension TodayViewController: NCWidgetProviding {
    
    func widgetPerformUpdateWithCompletionHandler(completionHandler: ((NCUpdateResult) -> Void)) {
        completionHandler(NCUpdateResult.newData)
    }
    
    func widgetMarginInsetsForProposedMarginInsets(defaultMarginInsets: UIEdgeInsets) -> UIEdgeInsets {
        var defaultMarginInsetsResult = defaultMarginInsets
        defaultMarginInsetsResult.right = defaultMarginInsets.left
        view.needsUpdateConstraints()
        tableView?.needsUpdateConstraints()
        return defaultMarginInsetsResult
    }

}
